eth2/proto_array_fork_choice/src/proto_array_fork_choice.rs

```diff
@@ -0,0 +1,699 @@
+use crate::error::Error;
+use crate::proto_array::ProtoArray;
+use crate::ssz_container::SszContainer;
+use parking_lot::RwLock;
+use ssz::{Decode, Encode};
+use ssz_derive::{Decode, Encode};
+use std::collections::HashMap;
+use types::{Epoch, Hash256, Slot};
+
+pub const DEFAULT_PRUNE_THRESHOLD: usize = 256;
+
+#[derive(Default, PartialEq, Clone, Encode, Decode)]
+pub struct VoteTracker {
+    current_root: Hash256,
+    next_root: Hash256,
+    next_epoch: Epoch,
+}
+
+/// A Vec-wrapper which will grow to match any request.
+///
+/// E.g., a `get` or `insert` to an out-of-bounds element will cause the Vec to grow (using
+/// Default) to the smallest size required to fulfill the request.
+#[derive(Default, Clone, Debug, PartialEq)]
+pub struct ElasticList<T>(pub Vec<T>);
+
+impl<T> ElasticList<T>
+where
+    T: Default,
+{
+    fn ensure(&mut self, i: usize) {
+        if self.0.len() <= i {
+            self.0.resize_with(i + 1, Default::default);
+        }
+    }
+
+    pub fn get(&mut self, i: usize) -> &T {
+        self.ensure(i);
+        &self.0[i]
+    }
+
+    pub fn get_mut(&mut self, i: usize) -> &mut T {
+        self.ensure(i);
+        &mut self.0[i]
+    }
+
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = &mut T> {
+        self.0.iter_mut()
+    }
+}
+
+pub struct ProtoArrayForkChoice {
+    pub(crate) proto_array: RwLock<ProtoArray>,
+    pub(crate) votes: RwLock<ElasticList<VoteTracker>>,
+    pub(crate) balances: RwLock<Vec<u64>>,
+}
+
+impl PartialEq for ProtoArrayForkChoice {
+    fn eq(&self, other: &Self) -> bool {
+        *self.proto_array.read() == *other.proto_array.read()
+            && *self.votes.read() == *other.votes.read()
+            && *self.balances.read() == *other.balances.read()
+    }
+}
+
+impl ProtoArrayForkChoice {
+    pub fn new(
+        finalized_block_slot: Slot,
+        justified_epoch: Epoch,
+        finalized_epoch: Epoch,
+        finalized_root: Hash256,
+    ) -> Result<Self, String> {
+        let mut proto_array = ProtoArray {
+            prune_threshold: DEFAULT_PRUNE_THRESHOLD,
+            justified_epoch,
+            finalized_epoch,
+            nodes: Vec::with_capacity(1),
+            indices: HashMap::with_capacity(1),
+        };
+
+        proto_array
+            .on_new_block(
+                finalized_block_slot,
+                finalized_root,
+                None,
+                justified_epoch,
+                finalized_epoch,
+            )
+            .map_err(|e| format!("Failed to add finalized block to proto_array: {:?}", e))?;
+
+        Ok(Self {
+            proto_array: RwLock::new(proto_array),
+            votes: RwLock::new(ElasticList::default()),
+            balances: RwLock::new(vec![]),
+        })
+    }
+
+    pub fn process_attestation(
+        &self,
+        validator_index: usize,
```

Comment:
I find `validator_indices` more helpful here, it convey a better message (ie. all the indices that voted for this attestation`

Reply:
I generally prefer to keep the loops as high as possible, because if you accept slices to functions like this you'll need to do an unnecessary heap allocation if you have one index. It's also nice if you happen to already be iterating through validator indices you can just call this function, instead of having to do another iteration. Probably insignificant, that's just generally my guiding principle.

Rust does a bunch of in-lining during compilation so I suspect there won't be an overhead by calling `process_attestation` multiple times from a higher-level loop.

Reply:
Makes sense. That's great insight. Thank you